### DISCUSSION TITLE
Tasmota and pn532

### DISCUSSION THREAD

**Discussion by joecas65** [arendst/Tasmota#17928] [Q&A]
Hello, i have a pn532 rfid card reader connected to a nodemcu with latest version of tasmota-sensors firmware.
The pn532 can read uid cards, but can't write.
Command sensors40 s,XXXX returns always an error.
Thanks for your help

**Comment by Noschvie** [comment]
Relevant infos to help and investigate are missing. Which Tasmota firmware used? Complete logs which shows what happened.

**Reply by joecas65** [reply]
firmware: 12.3.1(sensors)
log:
`2:57:51.442 RSL: STATE = {"Time":"2023-02-10T12:57:51","Uptime":"0T00:10:11","UptimeSec":611,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":37,"MqttCount":0,"Wifi":{"AP":1,"SSId":"Vodafone2.4GHz-33927832","BSSId":"08:16:05:0B:BC:FD","Channel":12,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:05"}}
12:58:49.905 CMD: Sensor40 S,ILOVETASMOTA
12:58:49.910 RSL: RESULT = {"Command":"Error"}`

**Comment by Noschvie** [comment]
Seems you need to "self compile" with `#define USE_PN532_DATA_FUNCTION`
```
#define USE_PN532_HSU                          // Add support for PN532 using HSU (Serial) interface (+1k8 code, 140 bytes mem)
#define USE_PN532_DATA_FUNCTION                // Add sensor40 command support for erase, setting data block content (+1k7 code, 388 bytes mem)
#define USE_PN532_DATA_RAW                     // Allow DATA block to be used by non-alpha-numberic data (+ 80 bytes code, 48 bytes ram)
```

**Reply by joecas65** [reply]
ok i self compiled with that code.
Now i get:

`17:04:49.438 RSL: STATE = {"Time":"2023-02-10T17:04:49","Uptime":"0T00:00:09","UptimeSec":9,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":37,"MqttCount":0,"Wifi":{"AP":1,"SSId":"Vodafone2.4GHz-33927832","BSSId":"08:16:05:0B:BC:FD","Channel":12,"Mode":"11n","RSSI":50,"Signal":-75,"LinkCount":1,"Downtime":"0T00:00:04"}}
17:06:34.362 CMD: Sensor40 S,ILOVETASMOTA
17:06:34.365 NFC: PN532 - Next scanned tag ���?(
17:06:34.367 RSL: RESULT = {"Command":"Error"}`

**Reply by barbudor** [reply]
~~If the sensor is working in Read but Sensor 40 is not working, it is that your compile is wrong~~

**Reply by barbudor** [reply]
@joecas65  Apologies for my previous messages, you are not at fault

4 months ago @md5sum-as changed the syntax of the commands but doesn't seem to have updated the doc

The new commands are now:
- ERASE instead of E
- WRITE instead of S
There are also other new undocumented commands for NTAG21x and NTAG20x cards
- AUTH
- SET_PWD
- UNSET_PWD

**Reply by md5sum-as** [reply]
Oops. I am SORRY! . I will update the doc in the next 2-3 days.

**Reply by joecas65** [reply]
Thanks a lot barbudor. Now it works like a charm

**Reply by barbudor** [reply]
Happy to hear that :+1:

**Reply by drtorchwood** [reply]
Beyond the changes of `Sensor40` the `#define USE_PN532_DATA_RAW` is also outdated in the documentation. And the PN532 is completely missing in the commands overview (https://tasmota.github.io/docs/Commands/). 
Perhaps, a list of supported tags could be added. I had mistakenly bought some mifare ultralight a few years ago that did - at that time - not work for writing data. Now, I updated to the latest Tasmota version and they seem to be compatible. 👍 The different versions of nfc tags is really confusing to me.

**Reply by barbudor** [reply]
@md5sum-as Any chance you could update the docs?
@drtorchwood if you feel the mood you can also propose doc update using the pencil icon on the top right of the page

**Reply by Noschvie** [reply]
@drtorchwood Have you tested the WRITE command with the mifare ultralight ?

**Reply by md5sum-as** [reply]
Pull request https://github.com/tasmota/docs/pull/1188
Awaiting for approval.

**Reply by barbudor** [reply]
Merged
Thanks for the doc update

**Reply by drtorchwood** [reply]
> @drtorchwood Have you tested the WRITE command with the mifare ultralight ?

I few year ago, I ordered some NFC tags that did not work at that time. I put a note "Mifare Ultralight" in the bag. If I recollect it correctly from the aliexpress order list, the description was "100pcs/Lot Ntag213 NFC TAG Sticker 13.56MHz ISO14443A NTAG 213 NFC tag Universal Lable for all NFC enabled phones". 
Now, with the new built I tried it again. The `Write` command worked for a string of 16 ASCII characters. 

I anybody knows how I can distinguish the different NFC tags, let me know. Perhaps, I will upload an older Tasmota version again to confirm that it did not work in the past.
